### PR TITLE
[Do not merge] Request for Comments: Example of Peer-to-Peer communication between tasks

### PR DESCRIPTION
**Pull request by rogancarr**
This branch contains an example of Peer-to-Peer communication between tasks, and contains a stand-alone communicator module that could be of use in other REEF projects where task-task communication is necessary.
 - Is this example worthwhile to add to REEF?
 - Is the Communicator module worth adding as a feature to REEF?

### CODE REVIEW

**Comment by rogancarr** [comment]
Checks:
- Travis failure is OpenJDK 7 -- normal now?
- AppVeyor 2017 failure is one test, looks like one of the normal failures
- AppVeyor 2015 failure is because of a C# coding style that isn't supported in 2015.
  - Will fix after review.

File: lang/cs/Org.Apache.REEF.Examples.PeerToPeer/Communication/MessageStreamingCodec.cs [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+        /// <param name="obj">The message to write</param>
+        /// <param name="writer">The writer to which to write</param>
+        public void Write(Message<T> obj, IDataWriter writer)
+        {
+            byte[] encodedMetadata = GenerateMetaDataEncoding(obj);
+            byte[] encodedInt = BitConverter.GetBytes(encodedMetadata.Length);
+            byte[] totalEncoding = encodedInt.Concat(encodedMetadata).ToArray();
+            writer.Write(totalEncoding, 0, totalEncoding.Length);
+
+            _codec.Write(obj.Data, writer);
```
Comment (motus) [inline lang/cs/Org.Apache.REEF.Examples.PeerToPeer/Communication/MessageStreamingCodec.cs]: Can't we just use the codec for conversion of objects into bytes? e.g. simply write `_codec.Write(obj, writer)`?

File: lang/cs/Org.Apache.REEF.Examples.PeerToPeer/Task/PeerTask.cs [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+
+            // Do a circular lookup for the peer Id
+            var peerNumber = _taskNumber == numberOfFriends - 1 ? 0 : _taskNumber + 1;
+            _peerId = CreateTaskIdFromNumericId(peerNumber);
+
+            // Stash the taskName for later
+            _taskId = taskId;
+
+            // Establish communications
+            _messageQueue = new Queue<string>();
```
Comment (motus) [inline lang/cs/Org.Apache.REEF.Examples.PeerToPeer/Task/PeerTask.cs]: Move initialization up to the declaration

File: lang/cs/Org.Apache.REEF.Examples.PeerToPeer/Task/PeerTask.cs [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+            [Parameter(typeof(PeerConfiguration.NumberOfPeers))] int numberOfFriends,
+            Communicator<string> communicator)
+        {
+            if (!GetTaskId(taskId, out _taskNumber))
+            {
+                throw new FormatException(string.Format("Could not extract Id from {0}", taskId));
+            }
+
+            // Do a circular lookup for the peer Id
+            var peerNumber = _taskNumber == numberOfFriends - 1 ? 0 : _taskNumber + 1;
```
Comment (motus) [inline lang/cs/Org.Apache.REEF.Examples.PeerToPeer/Task/PeerTask.cs]: maybe something like `(_taskNumber + 1) % numberOfFriends`?

File: lang/cs/Org.Apache.REEF.Examples.PeerToPeer/Task/PeerTask.cs [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+            // Stash the taskName for later
+            _taskId = taskId;
+
+            // Establish communications
+            _messageQueue = new Queue<string>();
+            _communicator = communicator;
+            _communicator.RegisterTask(_messageQueue);
+
+            // Set up the control flow
+            _rng = new Random();
```
Comment (motus) [inline lang/cs/Org.Apache.REEF.Examples.PeerToPeer/Task/PeerTask.cs]: Move to the declaration

File: lang/cs/Org.Apache.REEF.Examples.PeerToPeer/Task/PeerTask.cs [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+            _done = false;
```
Comment (motus) [inline lang/cs/Org.Apache.REEF.Examples.PeerToPeer/Task/PeerTask.cs]: same as above

File: lang/cs/Org.Apache.REEF.Examples.PeerToPeer/Task/PeerTask.cs [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+
+            int sentCount = 0;
+
+            // Have the first task initialize the conversation
+            if (_taskNumber == 0)
+            {
+                Logger.Log(Level.Info, string.Format("I'm #{0}. Starting the communication.", _taskId));
+                var messageToSend = string.Format("Hi from {0}: {1}", _taskId, sentCount++);
+                _communicator.Send(_peerId, messageToSend);
+                Logger.Log(Level.Info, string.Format("Sending: {0}", messageToSend));
```
Comment (motus) [inline lang/cs/Org.Apache.REEF.Examples.PeerToPeer/Task/PeerTask.cs]: Don't use `string.Format()` for logging. do
```C#
Logger.Log(Level.Info, "Sending: {0}", messageToSend);
```

File: lang/cs/Org.Apache.REEF.Examples.PeerToPeer/Task/PeerTask.cs [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+            }
+
+            while (!IsDone() || _messageQueue.Count > 0)
+            {
+                while (_messageQueue.Count > 0)
+                {
+                    var message = _messageQueue.Dequeue();
+
+                    Logger.Log(Level.Info, string.Format("Received a message: \"{0}\"", message));
```
Comment (motus) [inline lang/cs/Org.Apache.REEF.Examples.PeerToPeer/Task/PeerTask.cs]: don't use `string.Format()`

File: lang/cs/Org.Apache.REEF.Examples.PeerToPeer/Task/PeerTask.cs [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+
+                    if (IsDone())
+                    {
+                        continue;
+                    }
+
+                    if (message == EndMessage)
+                    {
+                        Logger.Log(Level.Info, string.Format("I heard {0} and I'm leaving.", EndMessage));
```
Comment (motus) [inline lang/cs/Org.Apache.REEF.Examples.PeerToPeer/Task/PeerTask.cs]: don't use `string.Format()`

File: lang/cs/Org.Apache.REEF.Examples.PeerToPeer/Task/PeerTask.cs [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+                    {
+                        Logger.Log(Level.Info, "Saying goodbye.");
+                        _communicator.Send(_peerId, EndMessage);
+                        Done();
+                        continue;
+                    }
+
+                    var messageToSend = string.Format("Hi from #{0}: {1}", _taskId, sentCount++);
+                    _communicator.Send(_peerId, messageToSend);
+                    Logger.Log(Level.Info, string.Format("Sending: {0}", messageToSend));
```
Comment (motus) [inline lang/cs/Org.Apache.REEF.Examples.PeerToPeer/Task/PeerTask.cs]: don't use `string.Format()`

File: lang/cs/Org.Apache.REEF.Examples.PeerToPeer/Task/PeerTask.cs [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+            _communicator.RegisterTask(_messageQueue);
+
+            // Set up the control flow
+            _rng = new Random();
+            _done = false;
+        }
+
+        public byte[] Call(byte[] memento)
+        {
+            Logger.Log(Level.Info, string.Format("Hello, World! I'm #{0}. I will now communicate with #{1}.", _taskId, _peerId));
```
Comment (motus) [inline lang/cs/Org.Apache.REEF.Examples.PeerToPeer/Task/PeerTask.cs]: don't use `string.Format()`

File: lang/cs/Org.Apache.REEF.Examples.PeerToPeer/Task/PeerTask.cs [hunk 11]
```diff
@@ -0,0 +1,153 @@
...
+
+            int sentCount = 0;
+
+            // Have the first task initialize the conversation
+            if (_taskNumber == 0)
+            {
+                Logger.Log(Level.Info, string.Format("I'm #{0}. Starting the communication.", _taskId));
```
Comment (motus) [inline lang/cs/Org.Apache.REEF.Examples.PeerToPeer/Task/PeerTask.cs]: don't use `string.Format()`

**Comment by markusweimer** [comment]
@rogancarr I am closing this for now. feel free to re-open as needed.

**Pull request closed without merging**